edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+
+```rust
+let query = "select 'Just a string'";
+let query_res: String = client.query_required_single(query, &()).await?;
+let query_res_opt: Option<String> = client.query_single(query, &()).await?;
+```
+
+## Using the `Queryable` macro
+
+The easiest way to unpack an EdgeDB query result is the built-in `Queryable` macro from the `edgedb-derive` crate. This turns queries directly into Rust types without having to match on a `Value` (more on the `Value` enum in [its own section](#the-value-enum)), cast to JSON, etc.
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct QueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let query = "with account := 
+    (insert Account {
+      username := 'User1'
+    }),
+    select account {
+      username,
+      id
+    };";
+let as_queryable_account: QueryableAccount = client
+    .query_required_single(query, &())
+    .await?;
+```
+
+Note: Field order within the shape of the query matters when using the `Queryable` macro. In the example before, a query is done in the order `id, username` instead of `username, id` as defined in the struct:
+
+```rust
+let query = "with account := 
+    (insert Account {
+        username := 'User1'
+    }),
+    select account {
+      id,
+      username
+    };";
+let wrong_order: Result<QueryableAccount, _> = client
+    .query_required_single(query, &())
+    .await;
+assert!(
+    format!("{wrong_order:?}")
+    .contains(r#"WrongField { unexpected: "id", expected: "username" }"#);
+);
+```
+
+You can use [cargo expand](https://github.com/dtolnay/cargo-expand) with the nightly compiler to see the code generated by the Queryable macro, but the minimal example repo also contains a somewhat cleaned up version of the generated code [here](https://github.com/Dhghomon/edgedb_rust_client_examples/blob/master/src/lib.rs#L12).
+
+## Passing in arguments
+
+Technically arguments can be passed in through any correctly formatted string, but this can be fragile and quickly gets awkward. Note the double `{{` curly braces as well as `' '` around the `movie_title` string to keep the EdgeDB compiler from seeing two separate `movie` and `title` tokens:
+
+```rust
+let movie_title = "Nice movie";
+let release_year = 2023;
+let query = format!(
+    "with movie_title := '{movie_title}',
+year := {release_year},
+movie := (insert Movie {{
+    title := movie_title,
+    release_year := year
+}}),
+select movie {{
+    title,
+    release_year,
+    id
+}}");
+```
+
+Instead, you can pass in arguments as a tuple as the second argument in the client methods. This allows the query to be a directly typed single `&'static str` and the cast notation (`<str>`, `<int32>`) makes the types more clearly visible:
+
+```rust
+let arguments = ("Nice movie", 2023);
+let query = "with 
+movie_title := <str>$0,
+year := <int32>$1,
+movie := (insert Movie {
+  title := movie_title,
+  release_year := year
+})
```

Comment:
What is the goal of using `with` everywhere? Here `Movie { title := $0, year := $1 }` give enough context for the variables to figure out which mean what. So adding with only makes it less readable I think.

Reply:
Okay, changed to that.